packages/edit-site/src/components/edit-template-part-menu-button/index.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import {
+	store as blockEditorStore,
+	BlockSettingsMenuControls,
+} from '@wordpress/block-editor';
+import { store as coreStore } from '@wordpress/core-data';
+import { MenuItem } from '@wordpress/components';
+import { isTemplatePart } from '@wordpress/blocks';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { store as editSiteStore } from '../../store';
+
+export default function EditTemplatePartMenuButton() {
+	const selectedTemplatePartId = useSelect( ( select ) => {
+		const { getSelectedBlockClientId, getBlock } = select(
+			blockEditorStore
+		);
+		const selectedBlockClientId = getSelectedBlockClientId();
+		const block = getBlock( selectedBlockClientId );
+
+		if ( isTemplatePart( block ) ) {
+			const templateParts =
+				select( coreStore ).getEntityRecords(
+					'postType',
+					'wp_template_part'
+				) || [];
+
+			const templatePart = templateParts.find(
+				( part ) =>
+					part.theme === block.attributes.theme &&
+					part.slug === block.attributes.slug
+			);
```

Comment:
It feels a little bit weird to have to do this to find the id of the template part. Maybe we should make this a selector? Or is there other API for this?

Reply:
It looks like template part CPTs have a fixed `"$theme//$slug"` ID structure so this should do the trick:

```js
const templatePart = select( coreStore ).getEntityRecord(
	'postType',
	'wp_template_part',
	`${ block.attributes.theme }//${ block.attributes.slug }`
);
```

https://github.com/WordPress/gutenberg/blob/217ff420e542bbd4714673b63f79fd4bee518035/lib/full-site-editing/block-templates.php#L265

Reply:
That feels like an internal implementation though which I'd like to avoid relying on. Unless it's being documented somewhere 🤔 ?

Reply:
There's already precedent for doing this in the codebase (e.g. `TemplatePartEdit`) and it should be safe to rely on this as it's part of the public REST API. Turns out there's actually a documented helper we can use.

https://github.com/WordPress/gutenberg/blob/839f540849410f8811f7259e513513c56557fa0c/packages/block-library/src/template-part/edit/utils/create-template-part-id.js#L1-L10

Reply:
The helper is not exposed though. Could it be easier if we just expose the `templatePartId` in block's attributes? So that we can just compare by ids rather than have to compose them first.

Reply:
But it's redundant to have both `id` and `part`+`slug` attributes in the Template Part block, though, no? They will contain the exact same data.

I think it's okay to rely on the ID being in this format. It's not an implementation detail as it's exposed in the public REST API. There's precedent already because `TemplatePartEdit` does exactly what we're doing here: it builds an ID from `part` and `slug` and passes it to `getEntityRecord`.

By using `getEntityRecord` instead of `getEntityRecords` we're not unnecessarily fetching all template parts from the REST API.

Reply:
Oh right, we can use `getEntityRecord` directly! I'm not sure where to put that `createTemplatePartId` though, it seems less "data"-ly to be put in `core-data`. I just inlined it here and added a comment.